clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -3515,4 +3668,198 @@ void handleFailure(Throwable throwable) {
+
+        Call metadataCall = getMetadataCall(context, 
+            () -> KafkaAdminClient.this.getListOffsetsCalls(context, topicPartitionOffsets, futures));
+        runnable.call(metadataCall, nowMetadata);
+
+        return new ListOffsetsResult(new HashMap<>(futures));
+    }
+
+    private List<Call> getListOffsetsCalls(MetadataOperationContext<ListOffsetsResultInfo, ListOffsetsOptions> context,
+            Map<TopicPartition, OffsetSpec> topicPartitionOffsets,
+            Map<TopicPartition, KafkaFutureImpl<ListOffsetsResultInfo>> futures) {
+
+        List<Call> calls = new ArrayList<>();
+        // grouping topic partitions per leader
+        Map<Node, Map<TopicPartition, ListOffsetRequest.PartitionData>> leaders = new HashMap<>();
+
+        Map<TopicPartition, Optional<Integer>> leaderEpochs = new HashMap<>();
+        for (TopicMetadata topicMetadata : context.getTopicMetadata()) {
+            for (PartitionMetadata partitionMetadata : topicMetadata.partitionMetadata()) {
+                leaderEpochs.put(new TopicPartition(topicMetadata.topic(), partitionMetadata.partition()), partitionMetadata.leaderEpoch());
+            }
+        }
+        for (Map.Entry<TopicPartition, OffsetSpec> entry: topicPartitionOffsets.entrySet()) {
+
+            OffsetSpec offsetSpec = entry.getValue();
+            long offsetQuery = (offsetSpec instanceof TimestampSpec)
+                    ? ((TimestampSpec)offsetSpec).timestamp()
+                    : (offsetSpec instanceof OffsetSpec.EarliestSpec)
+                        ? ListOffsetRequest.EARLIEST_TIMESTAMP
+                        : ListOffsetRequest.LATEST_TIMESTAMP;
+            // avoid sending listOffsets request for topics with errors
+            if (!context.getErrors().containsKey(entry.getKey().topic())) {
+                Node node = context.getCluster().leaderFor(entry.getKey());
+                if (node != null) {
+                    if (!leaders.containsKey(node))
+                        leaders.put(node, new HashMap<>());
+                    leaders.get(node).put(entry.getKey(), new ListOffsetRequest.PartitionData(offsetQuery, leaderEpochs.get(entry.getKey())));
+                } else {
+                    KafkaFutureImpl<ListOffsetsResultInfo> future = futures.get(entry.getKey());
+                    future.completeExceptionally(Errors.LEADER_NOT_AVAILABLE.exception());
+                }
+            }
+        }
+
+        for (final Map.Entry<Node, Map<TopicPartition, ListOffsetRequest.PartitionData>> entry: leaders.entrySet()) {
+
+            final int brokerId = entry.getKey().id();
+
+            calls.add(new Call("listOffsets on broker " + brokerId, context.getDeadline(), new ConstantNodeIdProvider(brokerId)) {
+
+                @Override
+                ListOffsetRequest.Builder createRequest(int timeoutMs) {
+                    return ListOffsetRequest.Builder
+                            .forConsumer(true, context.getOptions().isolationLevel())
```

Comment:
I think `requireTimestamp` is only needed if we are not requesting the earliest or latest offset.

Reply:
Here we could have a mix of timestamp, latest and earliest. Even set the broker returns `-1` when not using a timestamp.